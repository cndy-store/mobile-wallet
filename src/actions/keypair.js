import { getItem, setItem, removeItem } from '../lib/keypairStorage';
import { StellarSdk } from '../lib/stellar';
const { Keypair, StrKey } = StellarSdk;

const STORAGE_KEY = 'STELLAR_SECRET';

export const KEYPAIR_STARTS_PROCESSING = 'KEYPAIR_STARTS_PROCESSING';
export const KEYPAIR_ERROR = 'KEYPAIR_ERROR';
export const KEYPAIR_LOAD = 'KEYPAIR_LOAD';
export const KEYPAIR_SAVE = 'KEYPAIR_SAVE';
export const KEYPAIR_REMOVE = 'KEYPAIR_REMOVE';

const chomp = str => str.replace(/\n|\r| /g, '');
const isValidSecret = secret => StrKey.isValidEd25519SecretSeed(secret);

export const keypairProcessing = bool => ({
  type: KEYPAIR_STARTS_PROCESSING,
  isProcessing: bool
});

export const keypairErrored = ({ error }) => ({
  type: KEYPAIR_ERROR,
  error
});

export const keypairLoadSucceded = ({ secret }) => {
  const keypair = secret ? Keypair.fromSecret(secret) : null;

  return {
    type: KEYPAIR_LOAD,
    keypair: keypair,
    error: null
  };
};

export const keypairSaveSucceded = ({ secret }) => {
  const keypair = Keypair.fromSecret(secret);

  return {
    type: KEYPAIR_SAVE,
    keypair: keypair,
    error: null
  };
};

export const keypairRemoveSucceded = () => ({
  type: KEYPAIR_REMOVE,
  keypair: null,
  error: null
});

export const loadKeypair = () => {
  return async dispatch => {
    dispatch(keypairProcessing(true));
    try {
      const secret = await getItem(STORAGE_KEY);
      dispatch(keypairProcessing(false));
      dispatch(keypairLoadSucceded({ secret }));
      return Promise.resolve({ secret });
    } catch (error) {
      dispatch(keypairProcessing(false));
      dispatch(keypairErrored({ error }));
      return Promise.reject({ error });
    }
  };
};

export const saveKeypair = secret => {
  secret = chomp(secret);

  return async dispatch => {
    dispatch(keypairProcessing(true));

    if (!isValidSecret(secret)) {
      const error = new Error('Invalid Secret');
      dispatch(keypairProcessing(false));
      dispatch(keypairErrored({ error }));
      return Promise.reject({ error });
    }

    try {
      await setItem(STORAGE_KEY, secret);
      dispatch(keypairProcessing(false));
      dispatch(keypairSaveSucceded({ secret }));
      return Promise.resolve({ secret });
    } catch (error) {
      dispatch(keypairProcessing(false));
      dispatch(keypairErrored({ error }));
      return Promise.reject({ error });
    }
  };
};

export const removeKeypair = () => {
  return async dispatch => {
    dispatch(keypairProcessing(true));

    try {
      await removeItem(STORAGE_KEY);
      dispatch(keypairProcessing(false));
      dispatch(keypairRemoveSucceded());
      return Promise.resolve();
    } catch (error) {
      dispatch(keypairProcessing(false));
      dispatch(keypairErrored({ error }));
      return Promise.reject({ error });
    }
  };
};
