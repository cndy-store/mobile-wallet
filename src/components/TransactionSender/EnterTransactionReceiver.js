import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { StatusBar, View } from 'react-native';
import {
  Icon,
  Body,
  Card,
  CardItem,
  Container,
  Form,
  Input,
  Label,
  Header,
  Item,
  Left,
  Right,
  Content,
  Button,
  Text,
  Title
} from 'native-base';
import { styles as s } from 'react-native-style-tachyons';
import FormError from '../../components/FormError';
import Dimensions from 'Dimensions';

const { width, height } = Dimensions.get('window');

class EnterTransactionReceiver extends Component {
  constructor(props) {
    super(props);

    this.state = {
      receiverInput: ''
    };

    this.updateReceiver = this.updateReceiver.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
  }

  updateReceiver(receiverInput) {
    this.setState({ receiverInput });
  }

  handleSubmit() {
    this.props.onSubmit({ receiverInput: this.state.receiverInput });
  }

  renderError() {
    if (!this.props.error) return null;

    return <FormError message={this.props.error} />;
  }

  render() {
    return (
      <View>
        <Card>
          <CardItem header>
            <Text>Enter Public Key of Receiver</Text>
          </CardItem>
          <CardItem>
            <View style={{ flex: 1 }}>
              <Form>
                <Item stackedLabel error={!!this.state.error}>
                  <Label>Receiver</Label>
                  <Input
                    autoFocus
                    value={this.state.receiverInput}
                    autoCorrect={false}
                    onChangeText={this.updateReceiver}
                  />
                </Item>
                {this.renderError()}
              </Form>
            </View>
          </CardItem>
          <CardItem>
            <Body>
              <Button block onPress={this.handleSubmit}>
                <Text>Next</Text>
              </Button>
            </Body>
          </CardItem>
        </Card>
      </View>
    );
  }
}

EnterTransactionReceiver.propTypes = {
  error: PropTypes.string,
  onSubmit: PropTypes.func.isRequired
};

export default EnterTransactionReceiver;
