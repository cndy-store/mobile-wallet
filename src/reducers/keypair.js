import {
  KEYPAIR_STARTS_PROCESSING,
  KEYPAIR_ERROR,
  KEYPAIR_LOAD,
  KEYPAIR_SAVE,
  KEYPAIR_REMOVE
} from '../actions/keypair';

const defaultState = {
  isProcessing: false,
  error: null,
  keypair: null
};

export default function keypair(state = defaultState, action) {
  switch (action.type) {
    case KEYPAIR_STARTS_PROCESSING:
      return { ...state, isProcessing: action.isProcessing };
    case KEYPAIR_ERROR:
      return {
        ...state,
        error: action.error,
        isProcessing: action.isProcessing
      };
    case KEYPAIR_LOAD:
    case KEYPAIR_SAVE:
    case KEYPAIR_REMOVE:
      return {
        ...state,
        keypair: action.keypair,
        error: action.error,
        isProcessing: action.isProcessing
      };
    default:
      return state;
  }
}
