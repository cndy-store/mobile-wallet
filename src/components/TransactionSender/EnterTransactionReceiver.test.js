import React from 'react';
import { render } from '../../__tests__/renderer';
import { Button, TextInput } from 'react-native';
import { publicKey } from '../../__tests__/fixtures/keypair';
import EnterTransactionReceiver from './EnterTransactionReceiver';

const onSubmit = jest.fn();

it('renders correctly without an error', () => {
  const { toJSON } = render(
    <EnterTransactionReceiver error={null} onSubmit={onSubmit} />
  );
  expect(toJSON()).toMatchSnapshot();
});

it('calls onSubmit with entered public key', () => {
  const { root } = render(
    <EnterTransactionReceiver error={null} onSubmit={onSubmit} />
  );

  const input = root.findByType(TextInput);
  input.props.onChangeText(publicKey);

  root.findByType(Button).props.onPress();
  expect(onSubmit).toHaveBeenCalledWith({ receiverInput: publicKey });
});

it('renders correctly with an error', () => {
  const { toJSON } = render(
    <EnterTransactionReceiver
      receiver={publicKey}
      error={'With an error'}
      onSubmit={onSubmit}
    />
  );
  expect(toJSON()).toMatchSnapshot();
});
