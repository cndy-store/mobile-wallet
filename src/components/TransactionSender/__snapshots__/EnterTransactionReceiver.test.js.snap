// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly with an error 1`] = `
<View>
  <View
    style={
      Object {
        "backgroundColor": "#fff",
        "borderBottomWidth": 0.5,
        "borderColor": "#ccc",
        "borderLeftWidth": 0.5,
        "borderRadius": 2,
        "borderRightWidth": 0.5,
        "borderTopWidth": 0.5,
        "elevation": 3,
        "flex": 1,
        "flexWrap": "nowrap",
        "marginBottom": 5,
        "marginLeft": 2,
        "marginRight": 2,
        "marginTop": 5,
        "shadowColor": "#000",
        "shadowOffset": Object {
          "height": 2,
          "width": 0,
        },
        "shadowOpacity": 0.1,
        "shadowRadius": 1.5,
      }
    }
    virtual={undefined}
  >
    <View
      header={true}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#fff",
          "borderBottomWidth": null,
          "borderRadius": 2,
          "flexDirection": "row",
          "paddingBottom": 15,
          "paddingLeft": 15,
          "paddingRight": 15,
          "paddingTop": 15,
        }
      }
      virtual={undefined}
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#000",
            "fontFamily": "System",
            "fontSize": 16,
            "fontWeight": "600",
          }
        }
        uppercase={false}
        virtual={undefined}
      >
        Enter Public Key of Receiver
      </Text>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#fff",
          "borderRadius": 2,
          "flexDirection": "row",
          "paddingBottom": 10,
          "paddingLeft": 15,
          "paddingRight": 15,
          "paddingTop": 10,
        }
      }
      virtual={undefined}
    >
      <View
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <View
          style={Object {}}
          virtual={undefined}
        >
          <View
            accessibilityComponentType={undefined}
            accessibilityLabel={undefined}
            accessibilityTraits={undefined}
            accessible={true}
            collapsable={undefined}
            hasTVPreferredFocus={undefined}
            hitSlop={undefined}
            isTVSelectable={true}
            nativeID={undefined}
            onLayout={undefined}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "transparent",
                "borderBottomWidth": 1,
                "borderColor": "#D9D5DC",
                "borderLeftWidth": 0,
                "borderRightWidth": 0,
                "borderTopWidth": 0,
                "flexDirection": null,
                "height": 65,
                "marginLeft": 15,
                "opacity": 1,
              }
            }
            testID={undefined}
            tvParallaxProperties={undefined}
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Object {
                  "alignSelf": "flex-start",
                  "color": "#575757",
                  "fontSize": 15,
                  "left": null,
                  "paddingLeft": null,
                  "paddingRight": 5,
                  "paddingTop": 5,
                  "position": null,
                  "right": null,
                  "top": 5,
                }
              }
              virtual={undefined}
            >
              Receiver
            </Text>
            <TextInput
              allowFontScaling={true}
              autoCorrect={false}
              autoFocus={true}
              editable={true}
              onChangeText={[Function]}
              placeholderTextColor="#575757"
              style={
                Array [
                  Object {
                    "alignSelf": "stretch",
                    "color": "#000",
                    "flex": 1,
                    "fontSize": 17,
                    "height": 50,
                    "lineHeight": 18,
                    "marginLeft": null,
                    "paddingLeft": null,
                    "paddingRight": 5,
                    "top": 1.5,
                    "width": null,
                  },
                  Object {
                    "fontSize": 27,
                  },
                ]
              }
              underlineColorAndroid="rgba(0,0,0,0)"
              value=""
              virtual={undefined}
            />
          </View>
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Object {
                "color": "#000",
                "fontFamily": "System",
                "fontSize": 16,
              }
            }
            uppercase={false}
            virtual={undefined}
          >
            With an error
          </Text>
        </View>
      </View>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#fff",
          "borderRadius": 2,
          "flexDirection": "row",
          "paddingBottom": 10,
          "paddingLeft": 15,
          "paddingRight": 15,
          "paddingTop": 10,
        }
      }
      virtual={undefined}
    >
      <View
        style={
          Object {
            "alignItems": "flex-start",
            "alignSelf": "stretch",
            "flex": 1,
          }
        }
        virtual={undefined}
      >
        <View
          accessibilityComponentType={undefined}
          accessibilityLabel={undefined}
          accessibilityTraits={undefined}
          accessible={true}
          collapsable={undefined}
          hasTVPreferredFocus={undefined}
          hitSlop={undefined}
          isTVSelectable={true}
          nativeID={undefined}
          onLayout={undefined}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "alignSelf": "stretch",
              "backgroundColor": "#007aff",
              "borderBottomWidth": null,
              "borderColor": "#007aff",
              "borderLeftWidth": null,
              "borderRadius": 5,
              "borderRightWidth": null,
              "borderTopWidth": null,
              "elevation": 2,
              "flexDirection": "row",
              "height": 45,
              "justifyContent": "center",
              "opacity": 1,
              "paddingBottom": 6,
              "paddingTop": 6,
            }
          }
          testID={undefined}
          tvParallaxProperties={undefined}
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Object {
                "backgroundColor": "transparent",
                "color": "#fff",
                "fontFamily": "System",
                "fontSize": 16.5,
                "lineHeight": 19,
                "marginLeft": 0,
                "marginRight": 0,
                "paddingLeft": 16,
                "paddingRight": 16,
              }
            }
            uppercase={false}
            virtual={undefined}
          >
            Next
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;

exports[`renders correctly without an error 1`] = `
<View>
  <View
    style={
      Object {
        "backgroundColor": "#fff",
        "borderBottomWidth": 0.5,
        "borderColor": "#ccc",
        "borderLeftWidth": 0.5,
        "borderRadius": 2,
        "borderRightWidth": 0.5,
        "borderTopWidth": 0.5,
        "elevation": 3,
        "flex": 1,
        "flexWrap": "nowrap",
        "marginBottom": 5,
        "marginLeft": 2,
        "marginRight": 2,
        "marginTop": 5,
        "shadowColor": "#000",
        "shadowOffset": Object {
          "height": 2,
          "width": 0,
        },
        "shadowOpacity": 0.1,
        "shadowRadius": 1.5,
      }
    }
    virtual={undefined}
  >
    <View
      header={true}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#fff",
          "borderBottomWidth": null,
          "borderRadius": 2,
          "flexDirection": "row",
          "paddingBottom": 15,
          "paddingLeft": 15,
          "paddingRight": 15,
          "paddingTop": 15,
        }
      }
      virtual={undefined}
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#000",
            "fontFamily": "System",
            "fontSize": 16,
            "fontWeight": "600",
          }
        }
        uppercase={false}
        virtual={undefined}
      >
        Enter Public Key of Receiver
      </Text>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#fff",
          "borderRadius": 2,
          "flexDirection": "row",
          "paddingBottom": 10,
          "paddingLeft": 15,
          "paddingRight": 15,
          "paddingTop": 10,
        }
      }
      virtual={undefined}
    >
      <View
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <View
          style={Object {}}
          virtual={undefined}
        >
          <View
            accessibilityComponentType={undefined}
            accessibilityLabel={undefined}
            accessibilityTraits={undefined}
            accessible={true}
            collapsable={undefined}
            hasTVPreferredFocus={undefined}
            hitSlop={undefined}
            isTVSelectable={true}
            nativeID={undefined}
            onLayout={undefined}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "transparent",
                "borderBottomWidth": 1,
                "borderColor": "#D9D5DC",
                "borderLeftWidth": 0,
                "borderRightWidth": 0,
                "borderTopWidth": 0,
                "flexDirection": null,
                "height": 65,
                "marginLeft": 15,
                "opacity": 1,
              }
            }
            testID={undefined}
            tvParallaxProperties={undefined}
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Object {
                  "alignSelf": "flex-start",
                  "color": "#575757",
                  "fontSize": 15,
                  "left": null,
                  "paddingLeft": null,
                  "paddingRight": 5,
                  "paddingTop": 5,
                  "position": null,
                  "right": null,
                  "top": 5,
                }
              }
              virtual={undefined}
            >
              Receiver
            </Text>
            <TextInput
              allowFontScaling={true}
              autoCorrect={false}
              autoFocus={true}
              editable={true}
              onChangeText={[Function]}
              placeholderTextColor="#575757"
              style={
                Array [
                  Object {
                    "alignSelf": "stretch",
                    "color": "#000",
                    "flex": 1,
                    "fontSize": 17,
                    "height": 50,
                    "lineHeight": 18,
                    "marginLeft": null,
                    "paddingLeft": null,
                    "paddingRight": 5,
                    "top": 1.5,
                    "width": null,
                  },
                  Object {
                    "fontSize": 27,
                  },
                ]
              }
              underlineColorAndroid="rgba(0,0,0,0)"
              value=""
              virtual={undefined}
            />
          </View>
        </View>
      </View>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#fff",
          "borderRadius": 2,
          "flexDirection": "row",
          "paddingBottom": 10,
          "paddingLeft": 15,
          "paddingRight": 15,
          "paddingTop": 10,
        }
      }
      virtual={undefined}
    >
      <View
        style={
          Object {
            "alignItems": "flex-start",
            "alignSelf": "stretch",
            "flex": 1,
          }
        }
        virtual={undefined}
      >
        <View
          accessibilityComponentType={undefined}
          accessibilityLabel={undefined}
          accessibilityTraits={undefined}
          accessible={true}
          collapsable={undefined}
          hasTVPreferredFocus={undefined}
          hitSlop={undefined}
          isTVSelectable={true}
          nativeID={undefined}
          onLayout={undefined}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "alignSelf": "stretch",
              "backgroundColor": "#007aff",
              "borderBottomWidth": null,
              "borderColor": "#007aff",
              "borderLeftWidth": null,
              "borderRadius": 5,
              "borderRightWidth": null,
              "borderTopWidth": null,
              "elevation": 2,
              "flexDirection": "row",
              "height": 45,
              "justifyContent": "center",
              "opacity": 1,
              "paddingBottom": 6,
              "paddingTop": 6,
            }
          }
          testID={undefined}
          tvParallaxProperties={undefined}
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Object {
                "backgroundColor": "transparent",
                "color": "#fff",
                "fontFamily": "System",
                "fontSize": 16.5,
                "lineHeight": 19,
                "marginLeft": 0,
                "marginRight": 0,
                "paddingLeft": 16,
                "paddingRight": 16,
              }
            }
            uppercase={false}
            virtual={undefined}
          >
            Next
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
